# Project Structure Rules

## Quick Reference
**When to use**: Organizing folders, naming files, structuring components and services

## Core Rules

### ğŸ“ **Folder Organization**

#### Frontend (portfolio.next/)
```
src/
â”œâ”€â”€ app/                    # Next.js 15 app router pages
â”‚   â”œâ”€â”€ admin/             # Admin-only pages
â”‚   â”œâ”€â”€ api/               # API routes
â”‚   â””â”€â”€ [dynamic]/         # Dynamic routes
â”œâ”€â”€ components/            # Reusable React components
â”‚   â”œâ”€â”€ admin/            # Admin-specific components
â”‚   â”œâ”€â”€ layouts/          # Layout components
â”‚   â””â”€â”€ [feature]/        # Feature-specific components
â”œâ”€â”€ context/              # React Context providers
â”œâ”€â”€ hooks/                # Custom React hooks
â”œâ”€â”€ lib/                  # Utility libraries and constants
â”œâ”€â”€ models/               # TypeScript interfaces and types
â”œâ”€â”€ services/             # API service functions
â””â”€â”€ utils/                # Helper utilities
```

#### Backend (portfolio.node/)
```
src/
â”œâ”€â”€ app.js                # Main application entry point
â”œâ”€â”€ config/               # Configuration files
â”œâ”€â”€ docs/                 # API documentation
â”œâ”€â”€ middlewares/          # Express middlewares
â”œâ”€â”€ models/               # Mongoose models
â”œâ”€â”€ routes/               # Express route handlers
â”‚   â”œâ”€â”€ admin/           # Admin-only routes
â”‚   â””â”€â”€ [feature]/       # Feature-specific routes
â”œâ”€â”€ services/            # Business logic services
â”œâ”€â”€ utils/               # Helper utilities
â””â”€â”€ validators/          # Input validation schemas
```

### ğŸ·ï¸ **Naming Conventions**

#### Files and Folders
- **PascalCase**: React components, TypeScript interfaces
- **camelCase**: JavaScript/TypeScript files, functions, variables
- **kebab-case**: URLs, API endpoints, CSS classes
- **UPPER_SNAKE_CASE**: Environment variables, constants

#### Examples
```typescript
// âœ… Good
components/AdminDashboard.tsx
services/userService.ts
models/User.ts
utils/logger.ts

// âŒ Bad
components/admin-dashboard.tsx
services/UserService.ts
models/user.ts
utils/Logger.ts
```

### ğŸ“„ **File Structure Patterns**

#### React Components
```typescript
// âœ… Standard component structure
'use client';

import { useState, useEffect } from 'react';
import { ComponentProps } from '@/models/Component';

interface Props {
  // Props interface
}

export default function ComponentName({ prop1, prop2 }: Props) {
  // Component logic
  return (
    // JSX
  );
}
```

#### Service Files
```typescript
// âœ… Service structure
import { API_ENDPOINTS } from '@/lib/constants';
import { ServiceType } from '@/models/Service';

export async function serviceFunction(): Promise<ServiceType> {
  try {
    // Service logic
  } catch (error) {
    // Error handling
  }
}
```

#### Backend Models
```javascript
// âœ… Mongoose model structure
const mongoose = require("mongoose");

const ModelSchema = new mongoose.Schema({
  // Schema definition
}, {
  timestamps: true
});

// Pre-save hooks
ModelSchema.pre("save", async function (next) {
  // Pre-save logic
});

module.exports = mongoose.model("ModelName", ModelSchema);
```

## Common Patterns

### ğŸ”„ **Import/Export Patterns**

#### Frontend Imports
```typescript
// âœ… Absolute imports with @ alias
import { Component } from '@/components/Component';
import { useHook } from '@/hooks/useHook';
import { Service } from '@/services/service';
import { Model } from '@/models/Model';

// âœ… Relative imports for closely related files
import { SubComponent } from './SubComponent';
import { types } from './types';
```

#### Backend Imports
```javascript
// âœ… Relative imports for internal modules
const service = require("../services/service");
const model = require("../models/Model");
const middleware = require("../middlewares/middleware");

// âœ… Package imports
const express = require("express");
const mongoose = require("mongoose");
```

### ğŸ¯ **Component Organization**

#### Feature-Based Structure
```
components/
â”œâ”€â”€ admin/
â”‚   â”œâ”€â”€ AdminDashboard.tsx
â”‚   â”œâ”€â”€ UserList.tsx
â”‚   â””â”€â”€ EditForm.tsx
â”œâ”€â”€ blog/
â”‚   â”œâ”€â”€ BlogView.tsx
â”‚   â””â”€â”€ BlogList.tsx
â””â”€â”€ project/
    â”œâ”€â”€ ProjectView.tsx
    â””â”€â”€ ProjectCard.tsx
```

#### Shared Components
```
components/
â”œâ”€â”€ Header.tsx
â”œâ”€â”€ Footer.tsx
â”œâ”€â”€ LoadingOverlay.tsx
â””â”€â”€ MarkdownEditor.tsx
```

## Quick Checklist

- [ ] **Folder structure** follows established patterns
- [ ] **File naming** uses correct conventions
- [ ] **Import paths** use absolute imports with @ alias (frontend)
- [ ] **Component structure** follows standard template
- [ ] **Service organization** separates business logic
- [ ] **Model definitions** include proper schemas and hooks
- [ ] **Route organization** groups by feature and access level
- [ ] **Type definitions** are in separate model files

## Cross-References

- [Frontend Rules](frontend-rules.md) - React component patterns
- [Backend Rules](backend-rules.md) - Express and MongoDB patterns
- [Service Patterns](service-patterns.md) - Business logic organization

---

*Follow these structure patterns to maintain consistency across the project.*
